docker/importer/importer.py

```diff
@@ -455,6 +476,98 @@ def thread_init():
+    """Process deletions from a GCS bucket source.
+
+    This validates the continued existence of every Bug in Datastore (for the
+    given source) against every bug currently in that source's GCS bucket,
+    calculating the delta. The bugs determined to have been
+    deleted from GCS are then flagged for deletion from Datastore by the worker.
+
+    If the delta is too large, something undesirable has been assumed to have
+    happened and deletion is aborted.
+    """
+
+    logging.info("Begin processing bucket for deletions: %s", source_repo.name)
+
+    # Get all the existing non-withdrawn Bug IDs for
+    # source_repo.name in Datastore
+    query = osv.Bug.query()
+    query = query.filter(osv.Bug.source == source_repo.name)
+    result = list(query.fetch(keys_only=False))
+    result.sort(key=lambda r: r.id())
+    vuln_ids_for_source = [
+        (r.id(), r.source_id.split(':')[1]) for r in result if not r.withdrawn
+    ]
+
+    storage_client = storage.Client()
+    # Get all of the existing records in the GCS bucket
+    # (to get their IDs for checking against Datastore)
+    listed_blobs = list(
+        storage_client.list_blobs(
+            source_repo.bucket, prefix=source_repo.directory_path))
+
+    import_failure_logs = []
+
+    def convert_blob_to_vuln(blob: storage.Blob) -> Optional[Tuple[str]]:
+      """Download and parse GCS blob into [vuln.id]"""
+      if not _is_vulnerability_file(source_repo, blob.name):
+        return None
+
+      # Use the _client_store thread local variable
+      # set in the thread pool initializer
+      # Download in a blob generation agnostic way to cope with the file
+      # changing between when it was listed and now.
+      blob_bytes = storage.Blob(
+          blob.name, blob.bucket,
+          generation=None).download_as_bytes(_client_store.storage_client)
+
+      try:
+        vulns = osv.parse_vulnerabilities_from_data(
+            blob_bytes,
+            os.path.splitext(blob.name)[1],
+            strict=self._strict_validation)
+        for vuln in vulns:
+          return vuln.id
+
+        return None
+      except Exception as e:
+        logging.error('Failed to parse vulnerability %s: %s', blob.name, e)
+        # Don't include error stack trace as that might leak sensitive info
+        # List.append() is atomic and threadsafe.
+        import_failure_logs.append('Failed to parse vulnerability "' +
+                                   blob.name + '"')
+        return None
+
+    # Setup storage client
+    def thread_init():
+      _client_store.storage_client = storage.Client()
+
+    with ThreadPoolExecutor(
+        _BUCKET_THREAD_COUNT, initializer=thread_init) as executor:
+      vulns_in_gcs = [
+          v for v in executor.map(convert_blob_to_vuln, listed_blobs) if v
+      ]
+
+      # diff
+      vulns_to_delete = list(
+          filter(lambda v: v[0] not in vulns_in_gcs, vuln_ids_for_source))
```

Comment:
per https://google.github.io/styleguide/pyguide.html#210-lambda-functions

Can we rewrite this as


```python
[v for v in vuln_ids_for_source if v[0] not in vulns_in_gcs]
```

Reply:
Done in e79c4b9